### PR TITLE
Rewrote call options to match current config

### PR DESCRIPTION
**Pull request by saicheems**
Simplifies the client config and implements retrying calls using
exponential backoff.

### CODE REVIEW

File: path_template.go [hunk 1]
```diff
@@ -1,276 +0,0 @@
-// Copyright 2016, Google Inc.
```
Comment (jmuk) [inline path_template.go]: I think removing path_template.go is not in the scope of this change, right?
This implementation is incomplete but probably it would be better to keep it until you refine / reimplement it.

**Comment by jmuk** [comment]
Good job, but I am wondering how this will unite with Go's idiomatic pattern of optional arguments. Usually several Go APIs provide options for varargs for With- prefixed constructors, and build actual option structs internally.  You'll see this pattern in [gcloud](https://godoc.org/google.golang.org/cloud) and [grpc](https://godoc.org/google.golang.org/grpc#DialOption). It's good to allow following this patterns.

File: path_template.go (hunk 1 above)
Comment (saicheems) [inline path_template.go]: Done.

**Comment by saicheems** [comment]
That's a good point.
Just pushed a change that brings the concept back, but allows the struct to be passed in by giving it a resolve method.

My reasoning for allowing the struct is so it can be generated easily by JSON.
PTAL and let me know if this is reasonable.

+ @vchudnov-g

File: gax.go [hunk 2]
```diff
@@ -1,191 +1,169 @@
...
+	// Configures the retry exponential backoff algorithm.
+	BackoffSettings BackoffSettings
+}
+
+// Parameters to the exponential backoff algorithm for retrying.
+type BackoffSettings struct {
+	// The initial delay time, in milliseconds, between the completion of
+	// the first failed request and the initiation of the first retrying
+	// request.
+	InitialRetryDelayMillis time.Duration
```
Comment (jmuk) [inline gax.go]: You should remove -Millis suffix, the unit is handled by time.Duration type.

File: api_callable.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			if !isRetryable(code) {
+				return nil, err
+			}
+			delayCtx, _ := context.WithTimeout(ctx, delay)
+			<-delayCtx.Done()
+
+			delay = shorter(scaleDuration(delay, backoffSettings.RetryDelayMultiplier),
+				backoffSettings.MaxRetryDelayMillis)
+			timeout = shorter(scaleDuration(timeout, backoffSettings.RPCTimeoutMultiplier),
+				backoffSettings.MaxRPCTimeoutMillis)
```
Comment (jmuk) [inline api_callable.go]: You don't have to check which is longer for delay/timeout or max timeouts.
A context will finish with timeout if the parent context is timed-out, or it is timed-out (like this: http://play.golang.org/p/YJng0EZXnz -- child_ctx.Done() ends in 1 second).

Because total timeout is set at line 49, delayCtx and timeoutCtx never pass the total timeout. You can simply scale them with multiplier.

File: api_callable.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package gax
+
+import (
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"time"
```
Comment (vchudnov-g) [inline api_callable.go]: I suggest the convention of having imports from the standard library, a blank line, and then the other imports.

File: gax.go [hunk 5]
```diff
@@ -1,191 +1,169 @@
...
+	// The multiplier by which to increase the delay time between the
+	// completion of failed requests, and the initiation of the subsequent
+	// retrying request.
+	RetryDelayMultiplier float64
+	// The maximum delay time, in milliseconds, between requests. When this
+	// value is reached, `RetryDelayMultiplier` will no longer be used to
+	// increase delay time.
+	MaxRetryDelayMillis time.Duration
```
Comment (jmuk) [inline gax.go]: Are there specific reasons to flatten the settings?  I personally like previous design of defining initial/max/multiplier into a single struct and defining methods (well, I like this mostly because I designed it).

I don't think there's a strict reason to follow the exact structure of the setting structs, and in Go context, I believe splitting into a struct is clearer.

File: api_callable.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		codes.Aborted,
+	}
+)
+
+// Represents a GRPC call stub.
+type APICall func(context.Context, interface{}) (interface{}, error)
+
+// isRetryable returns true if c represents a retryable GRPC code.
+func isRetryable(c codes.Code) bool {
+	for _, code := range retryableGrpcCodes {
```
Comment (vchudnov-g) [inline api_callable.go]: if you make retryableGrpcCodes a map, then this function body becomes a two-liner and lookup O(1)

File: api_callable.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+		if c == code {
+			return true
+		}
+	}
+	return false
+}
+
+// shorter returns the shorter time.Duration between a and b.
```
Comment (vchudnov-g) [inline api_callable.go]: s/between/of/

File: api_callable.go (hunk 4 above)
Comment (jmuk) [inline api_callable.go]: I also suggest to run gofmt command. It can take care of those imports or minor style errors.

File: api_callable.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+func shorter(a time.Duration, b time.Duration) time.Duration {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// scaleDuration returns a copy of a * mult.
```
Comment (vchudnov-g) [inline api_callable.go]: "returns the product of a and mult"

File: api_callable.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+func scaleDuration(a time.Duration, mult float64) time.Duration {
+	ns := float64(a) * mult
+	return time.Duration(int64(ns))
+}
+
+// stubWithRetry returns a wrapper for stub with a exponential backoff retry
```
Comment (vchudnov-g) [inline api_callable.go]: s/a/an/

File: api_callable.go (hunk 4 above)
Comment (saicheems) [inline api_callable.go]: This code is already formatted, so no action needed

File: api_callable.go (hunk 3 above)
Comment (saicheems) [inline api_callable.go]: Ah yeah you're totally right.

These are different timeouts than the one applied to the parent context (49 refers to a total timeout, these two are retry delay and rpc timeout) ... but I can simplify the code by applying the timeouts to their respective contexts before the call and not bother with the `shorter` method.

Will do.

File: api_callable.go (hunk 4 above)
Comment (jmuk) [inline api_callable.go]: It seems that import is not managed by gofmt then?  See  https://github.com/golang/go/wiki/CodeReviewComments#imports for the recommended style (and it says another command, goimports, to fix the import styles...)

Comment (saicheems) [inline api_callable.go]: Oh yeah nice I didn't know about that command

File: gax.go (hunk 5 above)
Comment (saicheems) [inline gax.go]: I think you're right, I just spent a bit looking at the configuration and it seems every other language massages it into their own data structure anyway. I had thought they were 1:1 mappings.

Will modify.

File: api_callable_test.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"testing"
+	"time"
+)
+
+var (
+	callSettings = &CallSettings{0, RetrySettings{[]string{"default"}, BackoffSettings{
+		100 * time.Millisecond,
```
Comment (vchudnov-g) [inline api_callable_test.go]: I think labeling the fields here (in all three structs) makes the code more readable)

File: gax.go [hunk 11]
```diff
@@ -1,191 +1,169 @@
...
-func ToArray(ctx context.Context, streamable PageStreamable, opts ...CallOption) ([]interface{}, error) {
-	var result []interface{}
-	err := PageStream(ctx, streamable, func(element interface{}) bool {
-		result = append(result, element)
-		return true
-	}, opts...)
-	return result, err
+func (s *CallSettings) Resolve(c *CallSettings) {
+	c.Timeout = s.Timeout
+	c.RetrySettings = s.RetrySettings
```
Comment (vchudnov-g) [inline gax.go]: This is a shallow copy, so s.RetrySettings.RetryCodes and c.RetrySettings.RetryCodes will refer to the same array; is this what you want? I assume the codes are not likely to be modified per-call, but still, if we ever add another map or array to RetrySettings....

File: gax_test.go [hunk 12]
```diff
@@ -1,299 +1,53 @@
...
-			parent_context, _ := context.WithTimeout(context.Background(), time.Duration(1.5*float64(timeUnit)))
-			_, err := Invoke(parent_context, nil, waitFunc, testOptions...)
-			return err
+var (
+	expected = &CallSettings{
+		time.Second * 1,
+		RetrySettings{
+			[]string{"hello", "world"},
+			BackoffSettings{
+				time.Second * 2,
```
Comment (vchudnov-g) [inline gax_test.go]: field names

Comment (saicheems) [inline gax_test.go]: Doneish, since this is a test that checks the resolve behavior, and I removed this particular path, it's not useful to comment on names.

File: gax.go (hunk 11 above)
Comment (saicheems) [inline gax.go]: Obsolete, but good call - I originally had that as a pointer, and didn't notice after I changed it.

File: api_callable_test.go (hunk 10 above)
Comment (saicheems) [inline api_callable_test.go]: Done

File: api_callable.go (hunk 9 above)
Comment (saicheems) [inline api_callable.go]: Done

File: api_callable.go (hunk 8 above)
Comment (saicheems) [inline api_callable.go]: Done

File: api_callable.go (hunk 7 above)
Comment (saicheems) [inline api_callable.go]: Obsolete

File: api_callable.go (hunk 6 above)
Comment (saicheems) [inline api_callable.go]: Done, good call

File: gax.go (hunk 2 above)
Comment (saicheems) [inline gax.go]: Done

**Comment by saicheems** [comment]
So... in the process of writing this I think I gained all the insight that Jun already had writing options the first time.

It's more or less what it was before now. I had thought that including a full public struct with settings would allow the user to easily configure settings `json.Marshal`, but I realized that the configuration json is different enough that it makes more sense to add that functionality as a separate method and instead keep the config idiomatic.

File: api_callable.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+		return stub(childCtx, data)
+	}
+}
+
+// CreateAPICall returns a wrapper for stub governed by the values provided in
+// settings.
+func CreateAPICall(ctx context.Context, stub APICall, opts ...CallOption) APICall {
+	settings := &callSettings{}
+	callOptions(opts).Resolve(settings)
+	if settings.retrySettings.retryCodes != nil {
```
Comment (jmuk) [inline api_callable.go]: nit: might be good to check the length instead of comparing with nil --- there can be non-nil but empty array, while it's safe to invoke len with nil (see http://play.golang.org/p/3STon08e3T)

File: api_callable.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+				return nil, ctx.Err()
+			}
+			timeoutCtx, _ := context.WithTimeout(ctx, backoffSettings.rpcTimeoutSettings.max)
+			timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
+			resp, err = stub(timeoutCtx, req)
+			code := grpc.Code(err)
+			if code == codes.OK {
+				break
+			}
+			if !retryableGRPCCodes[code] {
```
Comment (jmuk) [inline api_callable.go]: not using retry_codes in the config -- are you going to do it in a different patch?

Comment (saicheems) [inline api_callable.go]: I'll do it now. Good call - thanks for the reminder

File: api_callable.go (hunk 13 above)
Comment (saicheems) [inline api_callable.go]: Done

**Comment by saicheems** [comment]
+ @jba

**Comment by saicheems** [comment]
So the `Invoke` function is a part of this library, and I'm changing its name to match the convention in other languages:

It's now `CreateApiCall`, and it returns a wrapped stub based on provided settings.
Jonathan made a good point that we could pass an error returning function:
var resp *GetShelfResponse
`err := gax.Invoke(ctx, func() error {
   var err error
   resp, err = client.client.GetShelf(ctx, req)
   return err
})`
and avoid casting. It'd probably have to be renamed back to `Invoke`, but I think it's good to implement in a follow-up.

File: api_callable.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+			// If the deadline is exceeded...
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
+			}
+			timeoutCtx, _ := context.WithTimeout(ctx, backoffSettings.rpcTimeoutSettings.max)
+			timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
+			resp, err = stub(timeoutCtx, req)
+			code := grpc.Code(err)
+			if code == codes.OK {
+				break
```
Comment (jba) [inline api_callable.go]: I'd write "return resp, err" here. It's clearer.

**Comment by saicheems** [comment]
Although one difficulty is that invoke modifies the stub's context... so the stub passed to `gax.Invoke` would also have to take a context. It would really be:
```
var resp *GetShelfResponse
err := gax.Invoke(parentCtx, func(childCtx) error {
    var err error
    resp, err = client.client.GetShelf(childCtx, req)
    return err
})
```
Not sure if that's as nice.

**Comment by jba** [comment]
Yes, I forgot about passing the context in. It's still nice because of type
safety: there is no runtime casting.

File: api_callable.go (hunk 15 above)
Comment (saicheems) [inline api_callable.go]: Done

**Comment by jmuk** [comment]
Right, context has to be passed around. I actually had a similar discussion with @pongad about passing req/resp as interface{} or hiding it inside of caller environment, and somehow I chose interface{} one.
I have no strong opinion here -- passing req/resp would look more like API-calls, but I think that's it.

File: api_callable.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+			delay = scaleDuration(delay, backoffSettings.delayTimeoutSettings.multiplier)
+			timeout = scaleDuration(timeout, backoffSettings.rpcTimeoutSettings.multiplier)
+		}
+		return
+	}
+}
+
+// stubWithTimeout returns a wrapper for stub with a timeout applied to its
+// context.
+func stubWithTimeout(ctx context.Context, stub APICall, timeout time.Duration) APICall {
```
Comment (jba) [inline api_callable.go]: ctx isn't used in this function

File: api_callable.go [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+// scaleDuration returns the product of a and mult.
+func scaleDuration(a time.Duration, mult float64) time.Duration {
+	ns := float64(a) * mult
+	return time.Duration(int64(ns))
+}
+
+// stubWithRetry returns a wrapper for stub with an exponential backoff retry
+// mechanism based on the values provided in retrySettings.
+func stubWithRetry(ctx context.Context, stub APICall, retrySettings retrySettings) APICall {
```
Comment (jba) [inline api_callable.go]: ctx isn't used in this function.

Comment (saicheems) [inline api_callable.go]: Good call, totally didn't notice that
Fixed

File: api_callable.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+	"google.golang.org/grpc/codes"
+)
+
+// Represents a GRPC call stub.
+type APICall func(context.Context, interface{}) (interface{}, error)
+
+// scaleDuration returns the product of a and mult.
+func scaleDuration(a time.Duration, mult float64) time.Duration {
+	ns := float64(a) * mult
+	return time.Duration(int64(ns))
```
Comment (pongad) [inline api_callable.go]: `time.Duration(ns)`
No need to cast it int64, Duration itself is already an int64.

File: api_callable.go [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+			code := grpc.Code(err)
+			if code == codes.OK {
+				return resp, err
+			}
+			if !retrySettings.retryCodes[code] {
+				return nil, err
+			}
+			delayCtx, _ := context.WithTimeout(ctx, backoffSettings.delayTimeoutSettings.max)
+			delayCtx, _ = context.WithTimeout(delayCtx, delay)
+			<-delayCtx.Done()
```
Comment (pongad) [inline api_callable.go]: Is this

select {
  case <-ctx.Done():
  case <-time.After(backoffSettings.delayTimeoutSettings.max):
  case <-time.After(delay):
}

It seems silly to create a throwaway context. Not a big deal though; your choice.

File: api_callable.go (hunk 18 above)
Comment (saicheems) [inline api_callable.go]: It is, but since the function returns time.Duration, the strict typing throws an error without a cast.

File: api_callable.go (hunk 19 above)
Comment (saicheems) [inline api_callable.go]: I'll stick with it for now... it can be revisited. I don't think it's that big a deal
Maybe making a helper would be the cleanest. Then I could pass all durations and let it pick the shortest.

File: api_callable.go (hunk 18 above)
Comment (pongad) [inline api_callable.go]: This seems to work OK. Is it somehow different in your installation? https://play.golang.org/p/YhfWnkzfYU

File: api_callable.go (hunk 19 above)
Comment (pongad) [inline api_callable.go]: sgtm

File: api_callable.go (hunk 18 above)
Comment (saicheems) [inline api_callable.go]: Oooooh whoops did it backwards.
Fixed. Good call

**Comment by pongad** [comment]
lgtm. Not sure if other reviewers have any more concerns.
